kong/cache_warmup.lua

```diff
@@ -0,0 +1,140 @@
+local utils = require "kong.tools.utils"
+
+
+local cache_warmup = {}
+
+
+local tostring = tostring
+local ipairs = ipairs
+local math = math
+local kong = kong
+local ngx = ngx
+
+
+function cache_warmup._mock_kong(mock_kong)
+  kong = mock_kong
+end
+
+
+local function warmup_dns(premature, hosts, count)
+  if premature then
+    return
+  end
+
+  ngx.log(ngx.NOTICE, "warming up DNS entries ...")
+
+  local start = ngx.now()
+
+  for i = 1, count do
+    kong.dns.toip(hosts[i])
+  end
+
+  local elapsed = math.floor((ngx.now() - start) * 1000)
+
+  ngx.log(ngx.NOTICE, "finished warming up DNS entries",
+                      "' into the cache (in ", tostring(elapsed), "ms)")
+end
+
+
+local function fail_cb()
+  error("this should never be called as L2 should already be warmed")
+end
+
+
+local function cache_warmup_single_entity(dao)
+  local entity_name = dao.schema.name
+
+  ngx.log(ngx.NOTICE, "Preloading '", entity_name, "' into the cache ...")
+
+  local start = ngx.now()
+
+  local hosts, names, host_count
+  if entity_name == "services" then
+    hosts = {}
+    names = {}
```

Comment:
`names` is a very bad name for this variable. I suggest replacing it with `visited_hosts`.